python/private/stage2_bootstrap_template.py

```diff
@@ -0,0 +1,501 @@
+    else:
+        # Deduplicate returns a generator, but we need a list after this.
+        python_path_entries = list(python_path_entries)
+
+    # We're emulating PYTHONPATH being set, so we insert at the start
+    # This isn't a great idea (it can shadow the stdlib), but is the historical
+    # behavior.
+    runfiles_envkey, runfiles_envvalue = RunfilesEnvvar(module_space)
+    if runfiles_envkey:
+        os.environ[runfiles_envkey] = runfiles_envvalue
+
+    main_filename = os.path.join(module_space, main_rel_path)
+    main_filename = GetWindowsPathWithUNCPrefix(main_filename)
+    assert os.path.exists(main_filename), (
+        "Cannot exec() %r: file not found." % main_filename
+    )
+    assert os.access(main_filename, os.R_OK), (
+        "Cannot exec() %r: file not readable." % main_filename
+    )
+
+    # COVERAGE_DIR is set if coverage is enabled and instrumentation is configured
+    # for something, though it could be another program executing this one or
+    # one executed by this one (e.g. an extension module).
+    if os.environ.get("COVERAGE_DIR"):
+        cov_tool = FindCoverageEntryPoint(module_space)
+        if cov_tool is None:
+            PrintVerboseCoverage(
+                "Coverage was enabled, but python coverage tool was not configured."
+            )
+        else:
+            # Inhibit infinite recursion:
+            if "PYTHON_COVERAGE" in os.environ:
+                del os.environ["PYTHON_COVERAGE"]
+
+            if not os.path.exists(cov_tool):
+                raise EnvironmentError(
+                    "Python coverage tool %r not found. "
+                    "Try running with VERBOSE_COVERAGE=1 to collect more information."
+                    % cov_tool
+                )
+
+            # coverage library expects sys.path[0] to contain the library, and replaces
+            # it with the directory of the program it starts. Our actual sys.path[0] is
+            # the runfiles directory, which must not be replaced.
+            # CoverageScript.do_execute() undoes this sys.path[0] setting.
+            #
+            # Update sys.path such that python finds the coverage package. The coverage
+            # entry point is coverage.coverage_main, so we need to do twice the dirname.
+            coverage_dir = os.path.dirname(os.path.dirname(cov_tool))
+            PrintVerbose("coverage: adding to sys.path:", coverage_dir)
+            python_path_entries.append(coverage_dir)
+            python_path_entries = Deduplicate(python_path_entries)
+    else:
+        cov_tool = None
+
+    sys.stdout.flush()
+    # TODO: Putting all the import paths at the front isn't good because it
+    # can interfere with the stdlib. But it emulates the historical behavior
+    # of using PYTHONPATH to add import paths.
+    # NOTE: The sys.path must be modified before coverage is imported/activated
+    sys.path[0:0] = python_path_entries
```

Comment:
Since we now know where the python interpreter is, maybe we should have `stdlib` first and only then the rest of the packages? That way we don't shadow the `stdlib` and we may be able to avoid some surprising behaviour.

I am curious if we should keep the old behaviour with the 2 stage bootstrap or make minor modifications since we know better about the environment we are going to be executing.

Reply:
Ok, I looked into this. It looks fine.

In the google runtime, it goes [runfiles, stdlib, third party deps], but without any explanation as to why. I suspect it's just historical. It's just simple prepend/append without any special logic.

I updated the code to append the path and CI is happy. The one exception is the safepath behavior -- that path is still prepended. I think that's fine and matches regular Python behavior better anyways.